syntax = "proto3";
package dio;
option go_package = "github.com/nisaral/dio/pkg/pb";

service Orchestrator {
  // Python workers call this to announce they are ready
  rpc RegisterWorker (RegisterRequest) returns (RegisterResponse);
}

message RegisterRequest {
  string worker_id = 1;
  string address = 2; // e.g., "localhost:50051"
  repeated string models = 3;
}

message RegisterResponse {
  bool success = 1;
}

// Service for the Go Manager to call the Python Worker
service InferenceWorker {
  rpc Predict (InferenceRequest) returns (InferenceResponse);
  rpc CheckHealth (google.protobuf.Empty) returns (google.protobuf.Empty);
}

message InferenceRequest {
  string model_id = 1;
  bytes data = 2; 
}

message InferenceResponse {
  bytes output = 1;
  float latency_ms = 2;
}